core/validator_manager.go

```diff
@@ -0,0 +1,137 @@
+func NewValidatorManager(log Logger) *ValidatorManager {
+	return &ValidatorManager{
+		quorumSize:            big.NewInt(0),
+		validatorsVotingPower: nil,
+		log:                   log,
+		vpLock:                &sync.RWMutex{},
+	}
+}
+
+// Init sets voting power and quorum size
+func (vm *ValidatorManager) Init(validatorsVotingPower map[string]*big.Int) error {
+	vm.vpLock.Lock()
+	defer vm.vpLock.Unlock()
+
+	totalVotingPower := calculateTotalVotingPower(validatorsVotingPower)
+	if totalVotingPower.Cmp(big.NewInt(0)) <= 0 {
+		return errVotingPowerNotCorrect
+	}
+
+	vm.validatorsVotingPower = validatorsVotingPower
+	vm.quorumSize = calculateQuorum(totalVotingPower)
+
+	return nil
+}
+
+// HasQuorum provides information on whether messages have reached the quorum
+func (vm *ValidatorManager) HasQuorum(addressMap map[string]struct{}) bool {
+	vm.vpLock.RLock()
+	defer vm.vpLock.RUnlock()
+
+	// if not initialized correctly return false
+	if vm.validatorsVotingPower == nil {
+		return false
+	}
+
+	messageVotePower := big.NewInt(0)
+
+	for from := range addressMap {
+		if vote, ok := vm.validatorsVotingPower[from]; ok {
+			messageVotePower.Add(messageVotePower, vote)
+		}
+	}
+
+	return messageVotePower.Cmp(vm.quorumSize) >= 0
+}
+
+// HasPrepareQuorum provides information on whether prepared messages have reached the quorum
+func (vm *ValidatorManager) HasPrepareQuorum(proposalMessage *proto.Message, msgs []*proto.Message) bool {
+	if proposalMessage == nil {
+		vm.log.Info("HasPrepareQuorum - proposalMessage is not set")
+
+		return false
+	}
+
+	proposerAddress := proposalMessage.From
+	sendersAddressesMap := map[string]struct{}{
+		string(proposerAddress): {},
+	}
+
+	for _, message := range msgs {
+		if bytes.Equal(message.From, proposerAddress) {
+			vm.log.Info("HasPrepareQuorum - proposer is among signers but it is not expected to be")
+
+			return false
+		}
+
+		sendersAddressesMap[string(message.From)] = struct{}{}
+	}
+
+	return vm.HasQuorum(sendersAddressesMap)
+}
+
+func calculateQuorum(totalVotingPower *big.Int) *big.Int {
+	quorum := new(big.Int).Mul(totalVotingPower, big.NewInt(2))
+
+	return bigIntDivCeil(quorum, big.NewInt(3))
+}
+
+func calculateTotalVotingPower(validatorsVotingPower map[string]*big.Int) *big.Int {
+	totalVotingPower := big.NewInt(0)
+	for _, validatorVotingPower := range validatorsVotingPower {
+		totalVotingPower = totalVotingPower.Add(totalVotingPower, validatorVotingPower)
+	}
+
+	return totalVotingPower
+}
+
+// bigIntDivCeil performs integer division and rounds given result to next bigger integer number
+// It is calculated using this formula result = (a + b - 1) / b
+func bigIntDivCeil(a, b *big.Int) *big.Int {
+	result := new(big.Int)
+
+	return result.Add(a, b).
+		Sub(result, big.NewInt(1)).
+		Div(result, b)
+}
+
+// ConvertMessageToAddressSet converts messages slice to addresses map
+func ConvertMessageToAddressSet(messages []*proto.Message) map[string]struct{} {
```

Comment:
Consider unexporting this function?

Reply:
fixed https://github.com/0xPolygon/go-ibft/pull/72/commits/f73123680449e1b37230f7bd351a80fc63464f98